abci/proposals.go

```diff
@@ -0,0 +1,152 @@
+		sdkmempool.Mempool
+
+		// Factory defines the API/functionality which is responsible for determining
+		// if a transaction is a bid transaction and how to extract relevant
+		// information from the transaction (bid, timeout, bidder, etc.).
+		auction.Factory
+
+		// VerifyTx is utilized to verify a bid transaction according to the preferences
+		// of the top of block lane.
+		VerifyTx(ctx sdk.Context, tx sdk.Tx) error
+
+		// ProcessLaneBasic is utilized to verify the rest of the proposal according to
+		// the preferences of the top of block lane. This is used to verify that no
+		ProcessLaneBasic(txs [][]byte) error
+	}
+
+	// ProposalHandler contains the functionality and handlers required to\
+	// process, validate and build blocks.
+	ProposalHandler struct {
+		prepareLanesHandler blockbuster.PrepareLanesHandler
+		processLanesHandler blockbuster.ProcessLanesHandler
+		tobLane             TOBLaneProposal
+		logger              log.Logger
+		txEncoder           sdk.TxEncoder
+		txDecoder           sdk.TxDecoder
+	}
+)
+
+// NewProposalHandler returns a ProposalHandler that contains the functionality and handlers
+// required to process, validate and build blocks.
+func NewProposalHandler(
+	lanes []blockbuster.Lane,
+	tobLane TOBLaneProposal,
+	logger log.Logger,
+	txEncoder sdk.TxEncoder,
+	txDecoder sdk.TxDecoder,
+) *ProposalHandler {
+	return &ProposalHandler{
+		prepareLanesHandler: abci.ChainPrepareLanes(lanes...),
+		processLanesHandler: abci.ChainProcessLanes(lanes...),
+		tobLane:             tobLane,
+		logger:              logger,
+		txEncoder:           txEncoder,
+		txDecoder:           txDecoder,
+	}
+}
+
+// PrepareProposalHandler returns the PrepareProposal ABCI handler that performs
+// top-of-block auctioning and general block proposal construction.
+func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req cometabci.RequestPrepareProposal) cometabci.ResponsePrepareProposal {
+		// Build the top of block portion of the proposal given the vote extensions
+		// from the previous block.
+		topOfBlock := h.BuildTOB(ctx, req.LocalLastCommit, req.MaxTxBytes)
+
+		// If information is unable to be marshaled, we return an empty proposal. This will
+		// cause another proposal to be generated after it is rejected in ProcessProposal.
+		lastCommitInfo, err := req.LocalLastCommit.Marshal()
+		if err != nil {
+			h.logger.Error("failed to marshal last commit info", "err", err)
+			return cometabci.ResponsePrepareProposal{Txs: nil}
+		}
+
+		auctionInfo := &AuctionInfo{
+			ExtendedCommitInfo: lastCommitInfo,
+			MaxTxBytes:         req.MaxTxBytes,
+			NumTxs:             uint64(len(topOfBlock.Txs)),
+		}
+
+		// Add the auction info and top of block transactions into the proposal.
+		auctionInfoBz, err := auctionInfo.Marshal()
+		if err != nil {
+			h.logger.Error("failed to marshal auction info", "err", err)
+			return cometabci.ResponsePrepareProposal{Txs: nil}
+		}
```

Comment:
I'm a little bit concerned about this code segment here. Reason being if the `LocalLastCommit` is malformatted, there will essentially be an endless number of rounds of consensus. Same goes for marshalling `AuctionInfo`. Do you think this is potentially problematic and if so what do you think is the best solution to make sure the network can stay alive and continue broadcasting blocks?